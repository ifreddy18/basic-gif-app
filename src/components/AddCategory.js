import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState('');

    const handleOnChange = (e) => {
        setInputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        // eslint-disable-next-line no-unused-expressions
        e.preventDefault(); // Previene la recarga del navegador
        
        if ( inputValue.trim().length > 2) {
            setCategories( categories => [inputValue, ...categories]);
            setInputValue('');
        }
    };

    return (
        <form onSubmit={ handleSubmit }>
            <p>{ inputValue }</p>
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleOnChange }
            />
            {/* <button type="submit">Add category</button> */}
        </form>
    );
};

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
};