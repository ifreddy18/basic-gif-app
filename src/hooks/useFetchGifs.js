import { useEffect, useState } from "react";
import { getFetchGifs } from '../helpers/getFetchGifs';


export const useFetchGifs = ( category ) => {

    // const [images, setImages] = useState([]);

    const [state, setState] = useState({
        data: [],
        loading: true,
    });

    useEffect(() => {
        getFetchGifs( category )
            .then( imgs => {
                setState({
                    data: imgs,
                    loading: false
                });
            });        
    }, [ category ]); 


    return state; // { data: [], loading: true }
};