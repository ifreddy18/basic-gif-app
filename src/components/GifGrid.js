import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {

    // const [images, setImages] = useState([]);

    const { data:images, loading } = useFetchGifs( category );

    // Hook que se dispara la primera vez y luego no se dispara hasta que el
    // valor de 'category' cambia
    // useEffect(() => {
    //     getFetchGifs( category ).then( setImages );        
    // }, [ category ]); 
    
    return (
        <>
            <h2>{ category }</h2>
               
            { 
                loading && 
                <p className="animate__animated animate__flash">
                    Cargando...
                </p> }

            <div className="card-grid">
                {
                    images.map( img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))
                }
            </div>
        </>
    );

};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
};