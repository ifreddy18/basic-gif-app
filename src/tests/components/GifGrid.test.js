import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');// Simula ejecucion y resp de useFetchGifs

describe('Pruebas en <GifGrid />', () => {

    const category = 'Black Clover';

    test('Se debe mostrar el componene correctamente', () => {
        // Simulacion de del return del custom hook useFetchGifs
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <GifGrid category={ category } /> );
        expect( wrapper ).toMatchSnapshot();
    })

    test('Debe de mostrar items cuando se cargan imagenes de useFetchGifs', () => {
        
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier-imagen.jpeg',
            title: 'Este es el titulo'
        }];
        
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow( <GifGrid category={ category } /> );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GifGridItem').length ).toBe(gifs.length);
    
    })
    

});