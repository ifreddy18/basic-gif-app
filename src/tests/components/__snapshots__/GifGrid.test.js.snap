// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> Debe de mostrar items cuando se cargan imagenes de useFetchGifs 1`] = `
<Fragment>
  <h2>
    Black Clover
  </h2>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Este es el titulo"
      url="https://localhost/cualquier-imagen.jpeg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en <GifGrid /> Se debe mostrar el componene correctamente 1`] = `
<Fragment>
  <h2>
    Black Clover
  </h2>
  <p
    className="animate__animated animate__flash"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
